### PR TITLE
Implement block filters API for the RPC mode.

### LINKED ISSUE #835: Web3 Filters do not work
Initial issue here , please take a look at the comments https://github.com/status-im/status-react/issues/2479

### PR DESCRIPTION
**Pull request by mandrigin**
Some operations (like deploying contracts, see issue #835) require web3 filter APIs to work.

We don't run LES in the RPC mode, and methods for creating/subscribing block filters aren't available on Infura anymore. This PR re-implements filters as a separate service and disables forwarding of filter creation to Infura.

APIs implemented:
`eth_newBlockFilter`
`eth_uninstallFilter`
`eth_getFilterChanges`

Closes #835

### CODE REVIEW

**Comment by adambabik** [comment]
@mandrigin is the title of this PR valid? I can see a filters implementation actually. Please also update the PR description in plain English how this PR solves issue #835.

**Comment by mandrigin** [comment]
@adambabik "block" is not a verb here, it's not that I blocked filters, is that I implemented block filters. Yeah, it looks confusing, I'll update the title.

**Comment by adambabik** [comment]
@mandrigin ah thanks! I did not know they are called "block filters" actually.

File: geth/api/backend.go [hunk 1]
```diff
@@ -129,7 +131,17 @@ func (b *StatusBackend) startNode(config *params.NodeConfig) (err error) {
...
+
+	if config.UpstreamConfig.Enabled {
+		rpcFilters := func(*gethnode.ServiceContext) (gethnode.Service, error) {
+			svc := rpcfilters.New(b.statusNode)
+			return svc, nil
+		}
+		additionalServices = append(additionalServices, rpcFilters)
+	}
+
+	if err = b.statusNode.Start(config, additionalServices...); err != nil {
```
Comment (adambabik) [inline geth/api/backend.go]: I know that I added this possibility to pass services to `Start()` but all our services are registered in `node.go` with `activate*` functions. We can keep this change this way and I will create an issue to refactor `node.go`. Basically, the idea would be to register services like this:
```go
services := []gethnode.ServiceConstructor{}
appendIf(condition, rpcFiltersServiceConstructor),
// ...
```

Sounds good?

File: services/rpcfilters/api.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	"github.com/pborman/uuid"
+)
+
+type filter struct {
+	hashes []common.Hash
+	done   chan bool
+}
+
+func (f *filter) AddHash(hash common.Hash) {
+	if f.hashes == nil {
```
Comment (adambabik) [inline services/rpcfilters/api.go]: You can do `append` on a `nil` slice. so `f.hashes = append(f.hashes, hash)` should work in both cases.

File: services/rpcfilters/api.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+		f.hashes = []common.Hash{hash}
+	} else {
+		f.hashes = append(f.hashes, hash)
+	}
+}
+
+func newFilter() *filter {
+	return &filter{
+		hashes: nil,
```
Comment (adambabik) [inline services/rpcfilters/api.go]: It's `nil` by default.

File: services/rpcfilters/api.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+	go func() {
+		id, s := api.event.Subscribe()
+		defer api.event.Unsubscribe(id)
+
+		for {
+			select {
+			case hash := <-s:
+				api.filtersMu.Lock()
+				f.AddHash(hash)
```
Comment (adambabik) [inline services/rpcfilters/api.go]: If `AddHash` is a public method, maybe it can do the locking?

File: services/rpcfilters/api.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/pborman/uuid"
+)
+
+type filter struct {
+	hashes []common.Hash
+	done   chan bool
```
Comment (adambabik) [inline services/rpcfilters/api.go]: Usually, it's `chan struct{}` if the channel is never sent a value but used for synchronization purposes.

File: services/rpcfilters/api.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	return found
+}
+
+// GetFilterChanges returns the logs for the filter with the given id since
+// last time it was called. This can be used for polling.
+//
+// https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_getfilterchanges
+func (api *PublicAPI) GetFilterChanges(id rpc.ID) (interface{}, error) {
```
Comment (adambabik) [inline services/rpcfilters/api.go]: Why this return value is not `[]common.Hash`?

File: services/rpcfilters/api.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+	api.filtersMu.Lock()
+	defer api.filtersMu.Unlock()
+
+	if f, found := api.filters[id]; found {
+		hashes := f.hashes
+		f.hashes = nil
+		return returnHashes(hashes), nil
+	}
+
+	return []interface{}{}, fmt.Errorf("filter not found")
```
Comment (adambabik) [inline services/rpcfilters/api.go]: In return value it's `interface{}` but here it's `[]interface{}{}`. It can be `nil`.

File: services/rpcfilters/api.go (hunk 6 above)
Comment (adambabik) [reply services/rpcfilters/api.go]: Also, it says `returns the logs for the filter with the given id since last time it was called`. But hashes to `f.hashes` are only appended and never removed. This means it will return older hashes as well. Is it ok?

File: services/rpcfilters/latest_block_changed_event.go [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+	if latestBlock.Number().Cmp(e.previousLatestBlock.Number()) <= 0 {
+		return
+	}
+
+	e.previousLatestBlock = latestBlock
+
+	e.sxMu.Lock()
+	defer e.sxMu.Unlock()
+
+	// TASK: try using MUX here
```
Comment (adambabik) [inline services/rpcfilters/latest_block_changed_event.go]: What is MUX?

File: services/rpcfilters/latest_block_changed_event.go [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+	for _, channel := range e.sx {
```
Comment (adambabik) [inline services/rpcfilters/latest_block_changed_event.go]: I may be wrong but it looks like we can get a channel here, then it's closed because of canceling the subscription and we start sending to a closed channel.

**Review by adambabik — COMMENTED** [review]
The first round for now.

File: services/rpcfilters/latest_block_changed_event.go (hunk 9 above)
Comment (adambabik) [reply services/rpcfilters/latest_block_changed_event.go]: Oh scratch that. I confused this with other channels. It's good.

File: services/rpcfilters/api.go (hunk 4 above)
Comment (mandrigin) [reply services/rpcfilters/api.go]: Yeah, I think I'll move these mutexes inside.

File: services/rpcfilters/api.go (hunk 6 above)
Comment (mandrigin) [reply services/rpcfilters/api.go]: they are removed. a few lines down

```
		hashes := f.hashes
		f.hashes = nil
```

File: services/rpcfilters/api.go (hunk 7 above)
Comment (mandrigin) [reply services/rpcfilters/api.go]: yeah, that's because the function declaration is copied from the original geth filters API and it can return different values. I'll fix that.

File: services/rpcfilters/latest_block_changed_event.go (hunk 8 above)
Comment (mandrigin) [reply services/rpcfilters/latest_block_changed_event.go]: ah, that's a leftover, I thought about geth's event package.
decided not to use one.

File: geth/api/backend.go (hunk 1 above)
Comment (mandrigin) [reply geth/api/backend.go]: yep, that sounds good!
I'll implement that pattern for this service.
why I chose this way of registering, is because otherwise, it is too cumbersome to pass the RPC client to node.go.
I think some other services might benefit of that too, less code to pass around.

**Comment by mandrigin** [comment]
@adambabik the first round is done :)

File: services/rpcfilters/latest_block_changed_event.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// latestBlockChangedEvent represents an event that one can subscribe to
+type latestBlockChangedEvent struct {
+	sx                  map[int]chan common.Hash
+	sxMu                sync.Mutex
```
Comment (dshulyak) [inline services/rpcfilters/latest_block_changed_event.go]: it is a bit confusing to have mutex below protected variable

File: services/rpcfilters/latest_block_changed_event.go [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+	defer e.sxMu.Unlock()
+
+	delete(e.sx, id)
+}
+
+func newLatestBlockChangedEvent(provider latestBlockProvider) *latestBlockChangedEvent {
+	return &latestBlockChangedEvent{
+		sx:           make(map[int]chan common.Hash),
+		provider:     provider,
+		tickerPeriod: 3 * time.Second,
```
Comment (dshulyak) [inline services/rpcfilters/latest_block_changed_event.go]: const ?

File: services/rpcfilters/latest_block_changed_event.go [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// latestBlockChangedEvent represents an event that one can subscribe to
+type latestBlockChangedEvent struct {
+	sx                  map[int]chan common.Hash
+	sxMu                sync.Mutex
+	previousLatestBlock blockInfo
```
Comment (dshulyak) [inline services/rpcfilters/latest_block_changed_event.go]: why not common.Hash?

File: services/rpcfilters/latest_block_changed_event.go [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// latestBlockChangedEvent represents an event that one can subscribe to
+type latestBlockChangedEvent struct {
+	sx                  map[int]chan common.Hash
```
Comment (dshulyak) [inline services/rpcfilters/latest_block_changed_event.go]: what is sx?

File: services/rpcfilters/latest_block_changed_event_test.go [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+
+	wg := sync.WaitGroup{}
+	wg.Add(3)
+
+	testFunc := func() {
+		testEvent(t, event, []common.Hash{hash})
+		wg.Done()
+	}
+
+	go testFunc()
```
Comment (dshulyak) [inline services/rpcfilters/latest_block_changed_event_test.go]: for loop?

File: services/rpcfilters/api.go [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+	api.filtersMu.Unlock()
+
+	if found {
+		close(f.done)
+	}
+
+	return found
+}
+
+// GetFilterChanges returns the logs for the filter with the given id since
```
Comment (dshulyak) [inline services/rpcfilters/api.go]: we don't return logs, right? we support only block hashes

https://github.com/ethereum/wiki/wiki/JavaScript-API#web3ethfilter

**Review by dshulyak — APPROVED** [review]
have only minor comments, two questions though:

1. we need to support only filters for blocks or there is a plan to have filters for other options too?

from https://github.com/ethereum/wiki/wiki/JavaScript-API#web3ethfilter

2. there is a pub-sub alternative for every filter method. they are not used by web3?

File: services/rpcfilters/latest_block_changed_event.go [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+	defer e.sxMu.Unlock()
+	return len(e.sx)
+}
+
+func (e *latestBlockChangedEvent) processLatestBlock(latestBlock blockInfo) {
+	e.latestBlockMu.Lock()
+	defer e.latestBlockMu.Unlock()
+
+	// if we received the same or an older block than we already have, ignore it.
+	if latestBlock.Number().Cmp(e.previousLatestBlock.Number()) <= 0 {
```
Comment (dshulyak) [inline services/rpcfilters/latest_block_changed_event.go:71]: it might be a bug. how original implementation behaves if chain reorgs? it is definitely possible that multiple blocks will be mined on the same chain height

for example:
100: aaa -> 101: aab
100: aaa -> 101: aac -> 102: aaz

**Comment by mandrigin** [comment]
@dshulyak 
1) yeah, I think we can get away with just block filters for now. they are blocking contract deployments
2) we don't use pubsub now, that would be a nice thing to fix, but not in this PR :)

File: services/rpcfilters/latest_block_changed_event.go (hunk 13 above)
Comment (mandrigin) [reply services/rpcfilters/latest_block_changed_event.go]: subscriptions

File: services/rpcfilters/api.go (hunk 15 above)
Comment (mandrigin) [reply services/rpcfilters/api.go]: good catch!

File: services/rpcfilters/latest_block_changed_event.go (hunk 16 above)
Comment (mandrigin) [reply services/rpcfilters/latest_block_changed_event.go:71]: that is caused by infura's load balancing.

**Comment by dshulyak** [comment]
> we don't use pubsub now, that would be a nice thing to fix, but not in this PR :)

i meant web3 library, it is not using subscriptions, right?

for example web3 1.0.0, allows to use them https://github.com/ethereum/web3.js/blob/v1.0.0-beta.34/packages/web3-eth/src/index.js#L382-L407. but afair we use lower version

**Comment by mandrigin** [comment]
not the version that we use now

File: services/rpcfilters/latest_block_changed_event_test.go (hunk 14 above)
Comment (mandrigin) [reply services/rpcfilters/latest_block_changed_event_test.go]: that is a simple way to have 3 subscribers at the same time to make sure the all of them received a hash

File: services/rpcfilters/latest_block_changed_event.go (hunk 16 above)
Comment (dshulyak) [reply services/rpcfilters/latest_block_changed_event.go:71]: no, chain reorganization happens because ethereum doesn't provide instant finality for a block

https://blog.ethereum.org/2015/08/08/chain-reorganisation-depth-expectations/

basicaly the block is final only after certain depth of the chain. last block can be overwritten very likely

Comment (mandrigin) [reply services/rpcfilters/latest_block_changed_event.go:71]: I'm talking about that specific case. That could be easily seen if you request the latest block by RPC.
If the chain is A1, A2, A3.
You will (most probably) receive something like: A1, A2, A1, A2, A3, A2, A3, A3, A3, A3...
That is not the overwrites, because it is the same blocks.

Comment (dshulyak) [reply services/rpcfilters/latest_block_changed_event.go:71]: but you will discard any overwrite with that logic, so for example you tracked that the last block is:

number = 101 : hash = aaa

but then chain reorg happens and :

number = 101 : hash = bbb -> number = 102 : hash = zzz

and the last one is actually becomes canonical chain because it turned out to be longer, but you will never notify user about block with hash bbb

Comment (mandrigin) [reply services/rpcfilters/latest_block_changed_event.go:71]: yeah, that is true, I can keep a capped list of a few hashes instead.
how big of a history you think is worth keeping?

Comment (mandrigin) [reply services/rpcfilters/latest_block_changed_event.go:71]: I can address that in a follow-up PR

File: services/rpcfilters/api.go [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+// https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_getfilterchanges
+func (api *PublicAPI) GetFilterChanges(id rpc.ID) ([]common.Hash, error) {
+	api.filtersMu.Lock()
+	defer api.filtersMu.Unlock()
+
+	if f, found := api.filters[id]; found {
+		return f.PopHashes(), nil
+	}
+
+	return []common.Hash{}, fmt.Errorf("filter not found")
```
Comment (adambabik) [inline services/rpcfilters/api.go]: Could be `nil` and error could be `errors.New()`

File: services/rpcfilters/api.go [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+// returnHashes is a helper that will return an empty hash array case the given hash array is nil,
+// otherwise the given hashes array is returned.
+func returnHashes(hashes []common.Hash) []common.Hash {
+	if hashes == nil {
+		return []common.Hash{}
```
Comment (adambabik) [inline services/rpcfilters/api.go:120]: `nil` should be fine as well.

File: services/rpcfilters/latest_block_changed_event.go [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+
+	e.sxMu.Lock()
+	defer e.sxMu.Unlock()
+
+	for _, channel := range e.sx {
+		channel <- e.previousLatestBlock.Hash
+	}
+}
+
+func (e *latestBlockChangedEvent) Stop() error {
```
Comment (adambabik) [inline services/rpcfilters/latest_block_changed_event.go]: Do we need to return `error`? It look like it's not possible.

File: services/rpcfilters/latest_block_changed_event_test.go (hunk 14 above)
Comment (adambabik) [reply services/rpcfilters/latest_block_changed_event_test.go]: You can tie it up with a variable and a loop:
```go
repeat := 3
wg.Add(repeat)
for i := 0; i < repeat; i++ {
  go testFunc()
}
```

**Review by adambabik — APPROVED** [review]

File: services/rpcfilters/api.go (hunk 18 above)
Comment (mandrigin) [reply services/rpcfilters/api.go:120]: The original (geth) implementation returns an empty array, so I decided to follow.

File: services/rpcfilters/api.go (hunk 17 above)
Comment (mandrigin) [reply services/rpcfilters/api.go]: The original implementation in geth returns an empty array in that case. I wanted to keep it consistent.

File: services/rpcfilters/latest_block_changed_event.go (hunk 19 above)
Comment (mandrigin) [reply services/rpcfilters/latest_block_changed_event.go]: That way it becomes a valid node service, but maybe we don't need it :)

Comment (adambabik) [reply services/rpcfilters/latest_block_changed_event.go]: Oh, then we probably do. It won't conform to the interface.

Comment (mandrigin) [reply services/rpcfilters/latest_block_changed_event.go]: I don't use it as a service, that is a leftover of my first attempt. Thanks for pointing me out. 👍

File: geth/api/backend.go [hunk 20]
```diff
@@ -423,3 +434,10 @@ func (b *StatusBackend) NotifyUsers(message string, payload fcmlib.NotificationP
 
 	return err
 }
+
+func appendIf(condition bool, services []gethnode.ServiceConstructor, service gethnode.ServiceConstructor) []gethnode.ServiceConstructor {
```
Comment (divan) [inline geth/api/backend.go:438]: Not sure if creating a new function worth just to shave off 2 lines of `if {}` block one time.

**Review by divan — APPROVED** [review]
Great stuff. 👍

File: geth/api/backend.go (hunk 20 above)
Comment (adambabik) [reply geth/api/backend.go:438]: The idea is to remove all `activateXYZService` functions and move the invocation here. So it will look like this:
```go
appendIf(conditionA, serviceConstructorA)
appendIf(conditionB, serviceConstructorB)
// ...
```
It should have more sense in this context.
